src/internal/log/loggers.go

```diff
@@ -190,6 +192,88 @@ func InitPachctlLogger() {
 	makeLoggerOnce(enc, os.Stderr, false, []zap.Option{zap.AddCaller()})
 }
 
+// InitBatchLogger creates a new logger for command-line tools that need to retain their logs on
+// error.  If the returned callback is called with no error, then the log file is deleted.  If it's
+// called with an error, the path to the log is printed, the log file is retained, and log.Exit is
+// called to log the error (and ensure everything is flushed).  If logFile is non-empty, then the
+// log file is always retained.
+func InitBatchLogger(logFile string) func(err error) {
+	cfg := minimalConsoleEncoder
+	if !color.NoColor { // Enable color if it's not disabled via flags.
+		cfg.EncodeLevel = zapcore.CapitalColorLevelEncoder
+	}
+	enc := zapcore.NewConsoleEncoder(cfg)
+	keepLog := logFile != ""
+	close := func() {}
+	makeLoggerOnce(enc, os.Stderr, false, []zap.Option{zap.AddCaller(), zap.WrapCore(
+		func(c zapcore.Core) zapcore.Core {
+			name := "batch"
+			if len(os.Args) > 0 {
+				name = filepath.Base(os.Args[0])
+			}
+			if logFile == "" {
+				var err error
+				// On error, out is set to "" again.
+				logFile, err = xdg.CacheFile(fmt.Sprintf("pachyderm/log/%s.%s.log", name, time.Now().In(time.UTC).Format("20060102T150405Z")))
```

Comment:
Are we hoping to support Windows filenames? If not, it would be more legible were this `time.RFC3339`. If so, then I guess we’re stuck — and ISO 8601 doesn’t allow mixing extended-format date & basic-format time.

Reply:
It’s fine as it is. `2006-01-02T15:04:05Z` _is_ so much more legible than `20060102T150405Z`. But it’s really no big deal.